lib/pbench/test/unit/server/query_apis/test_datasets_contents.py

```diff
@@ -0,0 +1,472 @@
+                        "_id": "9e95ccb385b7a7a2d70ededa07c391da",
+                        "_score": 0.0,
+                        "_source": {
+                            "parent": "/",
+                            "directory": "/1-default",
+                            "mtime": "2021-05-01T24:00:00",
+                            "mode": "0o755",
+                            "run_data_parent": "ece030bdgfkjasdkf7435e6a7a6be804",
+                            "authorization": {"owner": "1", "access": "private"},
+                            "@timestamp": "2021-05-01T24:00:00",
+                        },
+                    }
+                ],
+            },
+        }
+        index = self.build_index_from_metadata()
+
+        # get_expected_status() expects to read username and access from the
+        # JSON client payload, however this API acquires that information
+        # from the Dataset. Construct a fake payload corresponding to the
+        # attach_dataset fixture.
+        auth_json = {"user": "drb", "access": "private"}
+        expected_status = self.get_expected_status(
+            auth_json, build_auth_header["header_param"]
+        )
+
+        response = query_api(
+            self.pbench_endpoint,
+            self.elastic_endpoint,
+            self.payload,
+            index,
+            expected_status,
+            json=response_payload,
+            status=HTTPStatus.OK,
+            headers=build_auth_header["header"],
+        )
+        if expected_status == HTTPStatus.OK:
+            res_json = response.json
+            expected_result = {"directories": [], "files": []}
+            assert expected_result == res_json
+
+    def test_empty_query(
+        self,
+        server_config,
+        query_api,
+        pbench_token,
+        build_auth_header,
+        find_template,
+        provide_metadata,
+    ):
+        """
+        Check the API when a directory is empty.
+        """
+        response_payload = {
+            "took": 55,
+            "timed_out": False,
+            "_shards": {"total": 3, "successful": 3, "skipped": 0, "failed": 0},
+            "hits": {
+                "total": {"value": 0, "relation": "eq"},
+                "max_score": None,
+                "hits": [],
+            },
+        }
+        index = self.build_index_from_metadata()
+
+        # get_expected_status() expects to read username and access from the
+        # JSON client payload, however this API acquires that information
+        # from the Dataset. Construct a fake payload corresponding to the
+        # attach_dataset fixture.
+        auth_json = {"user": "drb", "access": "private"}
+        expected_status = self.get_expected_status(
+            auth_json, build_auth_header["header_param"]
+        )
+
+        response = query_api(
+            self.pbench_endpoint,
+            self.elastic_endpoint,
+            self.payload,
+            index,
+            expected_status,
```

Comment:
I wonder if it's this simple... although I don't think this explains `RequestsMock` complaining about the Elastic URI not being called. In this case we expect the `/datasets/contents` API to return `HTTPStatus.NOT_FOUND`, if the authenticated user doesn't preclude the query entirely. That is, if `self.get_expected_status` returns `HTTPStatus.OK` in this case I think we actually want to "expect" `HTTPStatus.NOT_FOUND`.

(We could make `get_expected_status` handle this a bit more cleanly by adding an optional parameter to return if the authentication looks OK rather than assuming `OK`, but for now just something like `expected_status if expected_status != HTTPStatus.OK else HTTPStatus.NOT_FOUND` might work).

Reply:
@riya-17 : Yes, you need to make this change:

```
        response = query_api(
            self.pbench_endpoint,
            self.elastic_endpoint,
            self.payload,
            index,
            expected_status if expected_status != HTTPStatus.OK else HTTPStatus.NOT_FOUND,
            json=response_payload,
            status=HTTPStatus.OK,
            headers=build_auth_header["header"],
        )
```

Reply:
I think the problem may be more of a challenge...I don't think we can call Flask's `abort()` in the context of the postprocessing function.  😭

Reply:
@riya-17, it looks like what we're supposed to do is raise `PostprocessError` -- here's some code from `pbench/server/api/resources/query_apis/datasets_detail.py`'s `postprocess()` function:
```
        if len(hits) == 0:
            raise PostprocessError(
                HTTPStatus.BAD_REQUEST, "The specified dataset has gone missing"
            )
```

Reply:
Sorry, I did see the usage of PostprocessError but I thought abort should work. I should have tried that, my bad :(